### PR TITLE
Room persistency for chat

### PR DESCRIPTION
**Pull request by OliHalu**
First approach to use room persistency for chat.

Problem with sending messages correctly

### CODE REVIEW

**Comment by kordianbruck** [comment]
@OliHalu please merge the latest master to your branch and resolve any conflicts.

File: app/src/main/java/de/tum/in/tumcampusapp/tumonline/TUMOnlineRequest.java [hunk 1]
```diff
@@ -177,6 +178,7 @@ public static boolean checkTokenInactive(Context c) {
                 tumManager.releaseLock(url);
             } catch (Exception e) {
                 Utils.log(e, "TUMonline request failed");
+                e.printStackTrace();
```
Comment (kordianbruck) [inline app/src/main/java/de/tum/in/tumcampusapp/tumonline/TUMOnlineRequest.java]: This should not be in here. Utils.log(e) logs the element to the console.

File: app/src/androidTest/java/de/tum/in/tumcampusapp/activities/CafeteriaActivityTest.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+public class CafeteriaActivityTest extends BaseActivityTest {
+
+    @Rule
+    public ActivityTestRule<CafeteriaActivity> mActivityRule = new ActivityTestRule<>(CafeteriaActivity.class);
+
+    @Override
+    public void mainComponentDisplayedTest() {
+        //TODO
+    }
+
```
Comment (pfent) [inline app/src/androidTest/java/de/tum/in/tumcampusapp/activities/CafeteriaActivityTest.java]: I think we can remove those empty tests

File: app/src/androidTest/java/de/tum/in/tumcampusapp/activities/CalendarActivityTest.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public class CalendarActivityTest extends BaseActivityTest {
+
+    @Rule
+    public ActivityTestRule<CalendarActivity> mActivityRule = new ActivityTestRule<>(CalendarActivity.class);
+
+    @Override
+    public void mainComponentDisplayedTest() {
+        // TODO
```
Comment (pfent) [inline app/src/androidTest/java/de/tum/in/tumcampusapp/activities/CalendarActivityTest.java]: Also empty

File: app/src/main/java/de/tum/in/tumcampusapp/adapters/ChatHistoryAdapter.java [hunk 4]
```diff
@@ -47,146 +43,141 @@ public void removeUnsent(ChatMessage msg) {
...
-            ChatMessage msg = ChatMessageManager.toObject(cursor);
-            outgoing = currentChatMember.getId() == msg.getMember()
-                                                       .getId();
+    public View getView(int position, View convertView, ViewGroup parent)   {
+        boolean outgoing;
+
+        if (getItemViewType(position)==0)   {
+            outgoing = true;
+        } else {
+            outgoing = false;
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/adapters/ChatHistoryAdapter.java]: simplyfy to:
`outgoing = getItemViewType(position)==0`

File: app/src/main/java/de/tum/in/tumcampusapp/adapters/ChatHistoryAdapter.java [hunk 5]
```diff
@@ -47,146 +43,141 @@ public void removeUnsent(ChatMessage msg) {
...
+        View listItem = convertView;
+        if (listItem == null)   {
+            listItem = LayoutInflater.from(mContext).inflate(layout, parent, false);
+            holder = new ViewHolder();
+            //Set UI elements
+            holder.layout = listItem.findViewById(R.id.chatMessageLayout);
+
+            holder.tvMessage = listItem.findViewById(R.id.tvMessage);
+            holder.tvTimestamp = listItem.findViewById(R.id.tvTime);
+            if(outgoing)    {
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/adapters/ChatHistoryAdapter.java]: that brace seems funky.
could you run the auto-formatter for all your committed files, please?

File: app/src/main/java/de/tum/in/tumcampusapp/database/dataAccessObjects/ChatMessageDao.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    @Query("SELECT c.* FROM chat_message c, (SELECT c1._id " +
+           "FROM chat_message c1 LEFT JOIN chat_message c2 ON c2._id=c1.previous " +
+           "WHERE c2._id IS NULL AND c1.room=:room " +
+           "ORDER BY c1._id DESC " +
+           "LIMIT 1) AS until " +
+           "WHERE c._id>=until._id AND c.room=:room " +
+           "ORDER BY c._id")
+    List<ChatMessage> getAll(int room);
+
+    @Query("UPDATE chat_message SET read=1 WHERE read=0 AND room=:room")
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/database/dataAccessObjects/ChatMessageDao.java]: I think the `read=0` condition is unnecessary...

File: app/src/main/java/de/tum/in/tumcampusapp/database/dataAccessObjects/ChatMessageDao.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+           "ORDER BY c._id")
+    Flowable<List<ChatMessage>> getUnreadListFlow(int room);
+
+    @Query("SELECT * FROM chat_message c, (SELECT c1._id " +
+           "FROM chat_message c1 LEFT JOIN chat_message c2 ON c2._id=c1.previous " +
+           "WHERE (c2._id IS NULL OR c1.read=1) AND c1.room=:room " +
+           "ORDER BY c1._id DESC " +
+           "LIMIT 1) AS until " +
+           "WHERE c._id>until._id AND c.room=:room " +
+           "ORDER BY c._id DESC " +
```
Comment (pfent) [inline app/src/main/java/de/tum/in/tumcampusapp/database/dataAccessObjects/ChatMessageDao.java]: All those queries smell like a lot of copy/pasta.
You can build part of the query in a constant string

**Comment by pfent** [comment]
that failing Jenkins test has nothing to do with this PR...
@kordianbruck can you merge this with the failing test?

**Pull request merged**